flygym/examples/game/controls.py

```diff
@@ -0,0 +1,321 @@
+            self.joystick_keys.clear()
+        return pjoystick_keys
+
+    def get_actions(self, state):
+        initiated_legs = np.zeros(6)
+        gain_left = 0
+        gain_right = 0
+
+        joystick_keys = self.retrieve_joystick_buttons()
+        if state == "single":
+            initiated_legs[list(joystick_keys)] = 1
+        elif state == "tripod":
+            if 2 in joystick_keys:  # center button (LH)
+                initiated_legs[0] = 1
+                initiated_legs[2] = 1
+                initiated_legs[4] = 1
+            if 5 in joystick_keys:
+                initiated_legs[1] = 1
+                initiated_legs[3] = 1
+                initiated_legs[5] = 1
+        elif state == "CPG":
+            ctrl_vector = self.retrieve_joystick_axis()
+            # get axis returns a value between -1 and 1
+            # The norm of control vector maps to the norm of [gain_right, gain_left]
+            # The axis 1 value, maps to the difference between gain_right and gain_left
+            normalized_norm = np.linalg.norm(ctrl_vector) / np.sqrt(2) * 1.2
+            gain_left = normalized_norm * (-1) * np.sign(ctrl_vector[1])
+            gain_right = normalized_norm * (-1) * np.sign(ctrl_vector[1])
+
+            offset_norm = np.abs(ctrl_vector[0]) * 0.6
+            if ctrl_vector[0] > 0:
+                # turn to the right: decrease right gain
+                gain_right -= offset_norm
+            elif ctrl_vector[0] < 0:
+                # turn to the left: decrease left gain
+                gain_left -= offset_norm
+        else:
+            raise ValueError("Invalid state")
+
+        return gain_left, gain_right, initiated_legs
+
+    def flush_keys(self):
+        with self.lock:
+            self.joystick_keys.clear()
+
+    def quit(self):
+        if not self.game_state.get_quit():
+            self.game_state.set_quit(True)
+        self.listener_thread.join()
+        self.flush_keys()
+        pygame.joystick.quit()
+        pygame.quit()
+        self.keyboard_listener.stop()
+        self.keyboard_listener.join()
+
+
+class KeyboardControl:
+    def __init__(self, game_state: GameState):
+        self.game_state = game_state
+        self.is_joystick = False
+
+        # Keyboard keys
+        self.CPG_keys = ["w", "s", "a", "d", "q"]
+        self.state_keys = ["p", "o", "i"]
+        self.leg_keys = ["t", "g", "b", "z", "h", "n"]
+        self.tripod_keys = ["g", "h"]
+
+        # Shared lists to store key presses
+        self.pressed_CPG_keys = []
+        self.pressed_state_keys = []
+        self.pressed_leg_keys = []
+        self.pressed_tripod_keys = []
+
+        self.lock = threading.Lock()
+
+        print("Starting key listener")
+        # Start the keyboard listener thread
+        self.listener = keyboard.Listener(on_press=self.on_press)
+        self.listener.start()
+
+    def on_press(self, key):
+        key_str = str(key)  # Gets the character of the key
```

Comment:
I had to change this line to `key_str = key.char if hasattr(key, 'char') else str(key)` to get any of my inputs to work.

I'm using `pynput 1.7.7` and `pygame 2.6.1`